### PR TITLE
[ADD] sql_shell: get easy access to database without having access to server

### PR DESCRIPTION
**Pull request by william-andre**

### CODE REVIEW

**Comment by legalsylvain** [comment]
Hi @william-andre 
Nice move !

If you want, you can take a look on some community modules developed some years ago :
https://github.com/OCA/reporting-engine/tree/16.0/sql_export (close with your implémentation)
https://github.com/OCA/reporting-engine/tree/16.0/bi_sql_editor (more technical, allowing to create materialized views to improve performance)

Note : in your current implementation, it is possible to call INSERT, UPDATE, TRUNCATE. Is it intended ?

(I'm available to talk with you if you want some feedbacks, talk about the design, etc...)

**Comment by ajepe** [comment]
> Hi @william-andre Nice move !
> 
> If you want, you can take a look on some community modules developed some years ago : https://github.com/OCA/reporting-engine/tree/16.0/sql_export (close with your implémentation) https://github.com/OCA/reporting-engine/tree/16.0/bi_sql_editor (more technical, allowing to create materialized views to improve performance)
> 
> Note : in your current implementation, it is possible to call INSERT, UPDATE, TRUNCATE. Is it intended ?
> 
> (I'm available to talk with you if you want some feedbacks, talk about the design, etc...)
The UI is looks better than the available options in the community

**Comment by william-andre** [comment]
The purpose is indeed to be able to execute any SQL queries, such as INSERT, UPDATE, DELETE FROM, but also EXPLAIN ANALYZE or CREATE INDEX.
It is not intended to be only reading data but also a debugging tool (doing a SELECT is actually the least useful feature of this module since it can already be done using a standard `read()` though the ORM)

**Comment by legalsylvain** [comment]
Hi @ajepe.

>  The UI is looks better than the available options in the community

Thanks for your quick review. Could you elaborate on this, though ? I don't get your point.

> The purpose is indeed to be able to execute any SQL queries, such as INSERT, UPDATE, DELETE FROM, but also EXPLAIN ANALYZE or CREATE INDEX.

I see. Thanks for the explanation.

> Even though some of these queried can already be done with a server action

Indeed, but the ORM "protects" data inconsistency. ``account_move.unlink()`` will fail if the account move is posted. ``pos_order_line.write()`` will faill if french certification is installed.
With this module, it is possible by the UI, to destroy the data, violating ACL, security checks and business checks. It is not a neutral change. 

In the other hand, I understand the need, that is valid. But as you propose the PR against a stable version, I think odoo SA should communicate on that topic. I'm sure some integrators will disable this module to prevent their client from accessing this functionality.

In other words, if today, someone would find the possibility to perform SQL write/read queries on the database via UI or xml-rpc, this would be considered as a severe security flaw. @rco-odoo, a point of view on this topic? 

Thanks.

**Comment by william-andre** [comment]
You can run any SQL query in a server action by running `env.cr.execute`. This has always been possible.
This module only makes it easier to get the response from the query.

**Comment by legalsylvain** [comment]
> You can run any SQL query in a server action by running env.cr.execute. This has always been possible.

You're right. Sorry for the noice, then.

File: addons/sql_shell/models/sql_shell.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    query = fields.Text()
+    result = fields.Json(readonly=True)
+
+    def run(self):
+        try:
+            with self.env.cr.savepoint():
+                self.env.cr.execute(self.query)
+                self.result = {
+                    'header': [d.name for d in self.env.cr.description],
+                    'rows': self.env.cr.fetchall(),
```
Comment (fmdl) [inline addons/sql_shell/models/sql_shell.py:17]: some request (update, delete) doesn't return rows

Comment (fmdl) [reply addons/sql_shell/models/sql_shell.py:17]: note :  maybe add a limit to prevent browser crash
`'rows': self.env.cr.fetchall()[:1000],`

Comment (william-andre) [reply addons/sql_shell/models/sql_shell.py:17]: > some request (update, delete) doesn't return rows

They return how many rows were updated, for instance

Comment (william-andre) [reply addons/sql_shell/models/sql_shell.py:17]: > note :  maybe add a limit to prevent browser crash
> `'rows': self.env.cr.fetchall()[:1000],`

This depends on the browser/computer. Use at your own risks.
The purpose is not to fetch a bajillion rows. The way it is (temporarily) currently implemented would make the server's memory explode too btw.

File: addons/web/static/lib/ace/mode-sql.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            }, {
+                token : "paren.lparen",
+                regex : "[\\(]"
+            }, {
+                token : "paren.rparen",
+                regex : "[\\)]"
+            }, {
+                token : "text",
+                regex : "\\s+"
+            } ]
```
Comment (Levizar) [inline addons/web/static/lib/ace/mode-sql.js:71]: if you want to avoid all those `"\\"` stuff, you can do `String(/\/)` that will convert it to `"\\"`

File: addons/web/static/lib/ace/mode-sql.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            "int|numeric|decimal|date|varchar|char|bigint|float|double|bit|binary|text|set|timestamp|" +
+            "money|real|number|integer"
+        );
+
+        var keywordMapper = this.createKeywordMapper({
+            "support.function": builtinFunctions,
+            "keyword": keywords,
+            "constant.language": builtinConstants,
+            "storage.type": dataTypes
+        }, "identifier", true);
```
Comment (Levizar) [inline addons/web/static/lib/ace/mode-sql.js:34]: If it's not gonna change, we should use `const` unless there is something I don't know about that is impacted.

**Review by Levizar — COMMENTED** [review]
my 2 cents when reading it

File: addons/web/static/lib/ace/mode-sql.js (hunk 3 above)
Comment (william-andre) [reply addons/web/static/lib/ace/mode-sql.js:34]: This file is mostly a copy paste from ACE, except some minor changes to integrate it in Odoo (such as the use off `require` etc)

File: addons/sql_shell/models/sql_shell.py (hunk 1 above)
Comment (william-andre) [reply addons/sql_shell/models/sql_shell.py]: My bad, I was too quick in testing, I fixed both remarks in the latest commit

File: addons/sql_shell/views/sql_shell_views.xml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    <record model="ir.actions.act_window" id="sql_shell_action">
+        <field name="name">SQL Shell</field>
+        <field name="type">ir.actions.act_window</field>
+        <field name="res_model">sql.shell</field>
+        <field name="target">new</field>
+        <field name="view_mode">form</field>
+    </record>
+
+    <menuitem name="SQL Shell" id="sql_shell_menu" parent="base.menu_custom" groups="base.group_no_one" action="sql_shell_action" sequence="10000"/>
```
Comment (legalsylvain) [inline addons/sql_shell/views/sql_shell_views.xml:33]:
```suggestion
    <menuitem name="SQL Shell" id="sql_shell_menu" parent="base.next_id_9" groups="base.group_no_one" action="sql_shell_action" sequence="100"/>
```
For the time being, this entry is set at the very last position of the technical menu, that doesn't make much sense.
Propose to replace : "Setting > Technical > SQL Shell" by "Setting > Technical > Database Structure > SQL Shell"

File: addons/sql_shell/models/sql_shell.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+from psycopg2 import Error as Psycopg2Error
+import json
+
+from odoo import models, fields, api
+
+class SQLShell(models.Model):
```
Comment (legalsylvain) [inline addons/sql_shell/models/sql_shell.py:6]: Naive question : Is there a reason to not use classic Transient Model like other wizard ?

**Comment by legalsylvain** [comment]
hi @william-andre. What is the state of this PR ? 

Some functional review. Something is wrong if the max rows is small. could you take a look ? 

Otherwise, LGTM.

**SELECT with max > 40 : OK**

screenshot-localhost_8016-2023 08 30-22_34_06

**SELECT with max < 40 : KO**

image

**EXPLAIN ANALYZE : OK** (nice feature, BTW)

image

image

**Destroy the database : OK**
(yes, I know, still possible by server action).

image

Then

image

**Comment by legalsylvain** [comment]
Hi @william-andre, thanks for your answer ! 

> The issue you raised with "Max Rows" is actually a protection to avoid crashing the browser and making too much traffic on the network; increase the limit at your own risk

No : take a look again on my prinscreen. I set 5 to max rows, and an error is raised and I can not see the result. That's not expected, don't you think ? 

> And this PR will probably never get accepted because of you TRUNCATE TABLE example. 

Well, indeed, there is no warning, but we could imagine quite simply to add a warning with some confirmation, if it is a "write" or a "delete" requests. (similar algorithm than the community module that prohibit the non select requests : https://github.com/OCA/reporting-engine/blob/16.0/sql_request_abstract/models/sql_request_mixin.py#L39)

> Even if it is already possible in a server action, this makes it a bit too easy to do...

well, go to "application > select ``product`` module > click on uninstall" will do basically similar "apocalypse by UI". 
IMO if correct warning is displayed, that is OK. Don't you think ?
